add a new FunctionSchema kind called scratch

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79659

The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

shunting314 committed:
add a new FunctionSchema kind called scratch

The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

[ghstack-poisoned]

shunting314 committed:
Update on "add a new FunctionSchema kind called scratch"


The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

[ghstack-poisoned]

shunting314 committed:
Update on "add a new FunctionSchema kind called scratch"


The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

[ghstack-poisoned]

shunting314 commented:
The CI failure does not seems to be related:
```
RuntimeError: 
Unknown custom class type c10d.ProcessGroup. Please ensure it is registered.:
c10d::broadcast(__torch__.torch.classes.c10d.ProcessGroup _0, Tensor[] _1, int _2, int _3, int _4) -> __torch__.torch.classes.c10d.Work _0
                                             ~~~~~~~~~~~~ <--- HERE
```

But I'll try rebase first since my trunk is a couple days of stale.

shunting314 committed:
Update on "add a new FunctionSchema kind called scratch"


The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

[ghstack-poisoned]

bdhirsh commented on torchgen/model.py:
```diff
@@ -1201,7 +1201,15 @@ def __post_init__(self) -> None:
+                    [
+                        arg
+                        for arg in self.arguments.out
+                        if not arg.name.startswith("_scratch_")
```
can we check the schema kind here? I'd rather bake in the invariant of `"_scratch_" -> SchemaKind.scratch` once, and just rely on the schema kind everywhere else.

shunting314 replied:
Hmm, but the code here does not only check if an op is an out variant or a scratch op. Be more specific,
1. if the op s a out variant op then `len(self.arguments.out)` should equal to `len(self.returns)`
2. if the op is a scratch op, then we should skip the _scratch_ tensor arguments when comparing the number of out arguments against the number of return items.

Case 1 is just handled naturally since it's a special case of 2.

Even if we check for SchemaKind here, we still need to filter out _scratch_ tensors in case 2 by checking the argument name.

Let me know if I misunderstood.

shunting314 replied:
So the reason here is _scratch_ tensors are not returned.

bdhirsh replied:
oh yep I misunderstood. Sounds good!

We might wanna make the `_scratch_` check more first-class, and introduce a method like `Argument.is_scratch()` (just so the string `_scratch_` isn't baked into the codegen in multiple places).

bdhirsh commented:
Just left light feedback, core changes look ok to me. Feel free to tag me on any subsequent codegen PR's that involve the scratch ops!

shunting314 commented on torchgen/model.py:
```diff
@@ -1293,7 +1304,7 @@ def kind(self) -> SchemaKind:
             return SchemaKind.inplace
         elif is_out:
-            return SchemaKind.out
+            return SchemaKind.scratch if is_scratch else SchemaKind.out
```
yep, I was thinking a flattened if statement impose some limitation that `elif is_scratch:` must comes before `elif isout:` .  But the assertion you added inspired me. I can add:
```
assert !is_scratch, "..."
```
in the block for `elif is_out:` to avoid misuse.

shunting314 committed:
Update on "add a new FunctionSchema kind called scratch"


The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

[ghstack-poisoned]

bdhirsh commented:
You'll also need to run `lintrunner -a` in OSS to fix the lint

shunting314 committed:
Update on "add a new FunctionSchema kind called scratch"


The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

[ghstack-poisoned]

shunting314 commented:
> You'll also need to run `lintrunner -a` in OSS to fix the lint

I'm developing on fbcode since most of my other changes are in fbcode. Thus so far I'm relying on GH to help me find the lint errors :) But thanks for the suggestions!

shunting314 committed:
Update on "add a new FunctionSchema kind called scratch"


The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

[ghstack-poisoned]

bdhirsh reviewed (approved):
lgtm!

shunting314 committed:
Update on "add a new FunctionSchema kind called scratch"


The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

[ghstack-poisoned]

shunting314 committed:
Update on "add a new FunctionSchema kind called scratch"


The scratch op expose intermetidate/scratch tensors used in kernel implementation
as kernel input arguments so a memory planning algorithm can plan memory for those
tensors.

Differential Revision: [D37194429](https://our.internmc.facebook.com/intern/diff/D37194429/)

[ghstack-poisoned]
